2020/7/luggage/src/main.rs

```diff
@@ -0,0 +1,202 @@
+use std::collections::HashSet;
+use std::env;
+use std::fs;
+use std::hash::{Hash, Hasher};
+
+fn main() -> Result<(), std::io::Error> {
+    let args : Vec<_> = env::args().collect();
+    let input = fs::read_to_string(&args[1])?;
+    let rules = Rules::parse(&input);
+
+    println!("shiny gold bags contain: {}", rules.contains_recursive("shiny gold"));
+
+    Ok(())
+}
+
+pub struct Rules<'a> {
+    rules: HashSet<Rule<'a>>,
```

Comment:
I used a map ([`HashMap`](https://doc.rust-lang.org/std/collections/struct.HashMap.html) would be the equivalent in Rust) keyed off the containing bag type in the rule.

Reply:
But then I also enumerated all possible bag types and walked down the graph rather than walking up the graph (which it looks like you're doing).

Reply:
I walked down as well.  I switched back and forth between HashSet and HashMap a couple of times.